ps_deployment/README.md

```diff
@@ -0,0 +1,121 @@
+		
+		* pros:
+			* separating code and image development - probably advantage in bigger projects where development and devops is separated
+		* cons:
+			* not copying .git to image, using default paths (/opt/app/src), maybe others which can probably require our scripts rework
+			
+	* normal image build (eg. `docker build ...`):
+		* pros:
+			* already in use in our environment
+			* no need to implement additional tool
+			* clean approach
+
+| ---------------------------------- | ----------- | ------------ |
+| image name                         | build time  | size         |
+| ---------------------------------- | ----------- | ------------ |
+| ps_normal_github_base              | 4m42.531s   | 797MB        |
+| ps_normal_github_final             | 0m12.925s   | 797MB        |
+| ---------------------------------- | ----------- | ------------ |
+| ps_normal_rpm_base                 | 5m8.026s    | 804MB        |
+| ps_normal_rpm_final                | 0m15.766s   | 804MB        |
+| ---------------------------------- | ----------- | ------------ |
+| ps_s2i_github_base                 | 4m38.620s   | 1.12GB       |
+| ps_s2i_github_final                | 0m10.715s   | 1.12GB       |
+| ---------------------------------- | ----------- | ------------ |
+
+### High level build flow strategies
+
+OpenShift vs. Public service
+	* OpenShift:
+		* pros:
+			* flexible
+			* no waiting in queue
+		* cons:
+			* additional maintenance
+			* additional resources = additional costs/failures because of resource limit
+	
+	* Public service:
+		* pros:
+			* our current approach - no radical changes required
+			* simple and straightforward
+			* free
+			
+		* cons:
+			* dependency on external service
+			* long queue time in case of service issue/high load
+
+#### OpenShift
+	* Source-to-Image -
+		 Source-to-Image (`sourceStrategy` in OC configuration) is using s2i tool for image deployment
+	
+	Links:
+	* [s2i details](https://docs.openshift.com/container-platform/3.11/architecture/core_concepts/builds_and_image_streams.html#source-build)
+	* [s2i strategy options](https://docs.openshift.com/container-platform/3.11/dev_guide/builds/build_strategies.html#source-to-image-strategy-options)
+	* [s2i tool blog](https://www.openshift.com/blog/create-s2i-builder-image)
+
+	* Image Stream - current configuration
+	* Custom Build - need more details
+
+## Conclusion
+
+	* Source-to-image (S2I) - no performance improvements, bigger image
+	* RPM vs github:
+		* from performance perspective basically same results 
+		* github repos should not have issues with OS version - will make deployment easier
+	* splitting image to base + final, will improve build times 
+	
+	AI:
```

Comment:
I would start with:

* creating a base image with Ansible to be shared by the service, worker and listener images (and any other image we have that needs it).
* breaking up the dependency between the `packit` and the worker images, and maybe even removing the `packit` image. I would say to install `packit` from source in the workers, from the `master` branch when building for stage and from `stable` when building for prod.
* Check the results and see if further layering or refinement of the build process is required.

What do you think?

Reply:
I agree completely, with one note:

>     * would say to install `packit` from source in the workers, from the `master` branch when building for stage and from `stable` when building for prod.

Or, as Franta pointed out, we [now](https://github.com/packit-service/packit/pull/879) should have [packit-master](https://copr.fedorainfracloud.org/coprs/packit/packit-master) & [packit-releases](https://copr.fedorainfracloud.org/coprs/packit/packit-releases) (doesn't exist yet, since no builds have been done yet) repos in Copr, which should contain the packit & ogr rpms for their respective `master` & `stable` (in `ogr` case it means the PyPI release) branches.

The benefit I see in installing RPMs instead of from git is that RPMs will take care of all the dependencies so we don't have to have them separately in some playbook.

Reply:
creating a base image with Ansible to be shared by the service, worker and listener images (and any other image we have that needs it).

https://projects.engineering.redhat.com/browse/PACKIT-443

    breaking up the dependency between the packit and the worker images, and maybe even removing the packit image. I would say to install packit from source in the workers, from the master branch when building for stage and from stable when building for prod.

https://projects.engineering.redhat.com/browse/PACKIT-444

    Check the results and see if further layering or refinement of the build process is required.
i think, yes - it will be essential, but will support it with data once card above are implemented

Reply:
> The benefit I see in installing RPMs instead of from git is that RPMs will take care of all the dependencies so we don't have to have them separately in some playbook.

Yes, this is a great benefit to have. On the other hand I don't see this approach to work when someone would need to build  the Packit Service images locally with some custom modifications in `packit` and/or `ogr`. Or am I missing something.

With `pip install` the above would be possible, we just need to expose the source for these packages as build arguments, defaulting to the GitHub repos, `ADD` these sources into the image, and `pip install` them.

I agree that maintaining the rpm dependencies in the playbooks is far from ideal. Could we maybe get a list of dependencies from `setup.cfg` and then install them with `dnf install 'python3dist(<name>)'`?

Reply:
> I don't see this approach to work when someone would need to build the Packit Service images locally with some custom modifications

agree, let's scratch my idea

> Could we maybe get a list of dependencies from `setup.cfg` and then install them with `dnf install 'python3dist(<name>)'`?

Not any straightforward I know of. Plus there are [exceptions (bodhi-client and koji required by packit and not in setup.cfg)](https://github.com/packit-service/packit/blob/master/setup.cfg#L59)